src/Sender.cpp

```diff
@@ -0,0 +1,42 @@
+#include "Sender.hpp"
+#include "NetworkSimulator.hpp"
+
+#include <iostream>
+
+Sender::Sender(std::uint64_t processing_delay_ns):
+        NetworkDevice(processing_delay_ns) {
+    id = NetworkDevice::last_given_device_id++;
+}
+
+void Sender::ProcessPacket(Packet p) {
+    Enqueue(p);
+    /* Next device in the flow path */
+    std::shared_ptr<Link> link = NextLink(p.m_destination_id);
+
+    /* Latency is a sum of processing and queuing delays */
+    std::uint64_t latency = processing_delay_per_packet + std::max<std::uint64_t>(0, completion_time - NetworkSimulator::Now());
+
+    std::uint64_t transmission_delay = 0;
+    /* If ACK flag set to false, send the packet to the next device */
+    if (!p.m_is_ack) {
+        /* Waiting for the link to process previous packets, then add distance (i.e. link transmission time) */
+        transmission_delay = (std::uint64_t)std::max<std::int64_t>(0, (std::int64_t)(link->last_process_time_ns - (NetworkSimulator::Now() + latency))) +
```

Comment:
The same applies here. Delegate queuing at the egress port and transmission to the link object.